lib/bamboo/view.ex

```diff
@@ -0,0 +1,201 @@
+defmodule Bamboo.View do
+  @moduledoc """
+  Compiles and renders templates defined in a given path.
+
+  Functions defined in the view are available to use in its templates.
+
+  ## Example
+
+      defmodule MyApp.EmailView do
+        use Bamboo.View, path: "/lib/my_app/email/templates"
+
+        def app_title do
+          "My Super App"
+        end
+      end
+
+      # lib/my_app/email_templates/welcome.html
+      <h1>Welcome to <%= app_title() %></h1>
+  """
+
+  defmodule UndefinedTemplateError do
+    @moduledoc """
+    Exception raised when a template cannot be found.
+    """
+    defexception [:module, :template, :available]
+
+    def message(exception) do
+      ~s"""
+      Could not render #{inspect(exception.template)} for #{inspect(exception.module)}
+
+      #{available_templates(exception.available)}
+      """
+    end
+
+    defp available_templates([]), do: "No templates were compiled for this module."
+
+    defp available_templates(available) do
+      "The following templates were compiled:\n\n" <>
+        Enum.map_join(available, "\n", &"* #{&1}") <>
+        "\n"
+    end
+  end
+
+  defmacro __using__(path: path) do
+    quote do
+      @templates_path unquote(path)
+      @before_compile unquote(__MODULE__)
+    end
+  end
+
+  defmacro __using__(opts) do
+    raise ArgumentError, """
+    expected Bamboo.View to have a path set, instead got: #{inspect(opts)}.
+
+    Please set a path where this view's template are defined e.g. use Bamboo.View, path: "/lib/my_app/email/templates/account"
+    """
+  end
+
+  defmacro __before_compile__(env) do
+    templates_path = Module.get_attribute(env.module, :templates_path)
+
+    results =
+      templates_path
+      |> templates_to_compile()
+      |> Enum.map(&template_and_name/1)
+      |> Enum.map(fn {template, name} ->
+        {name, compile(template, name)}
+      end)
+
+    template_names = Enum.map(results, &elem(&1, 0))
+    render_functions = Enum.map(results, &elem(&1, 1))
+
+    quote do
+      unquote(render_functions)
+
+      def render_template(template, _assigns) do
+        raise UndefinedTemplateError,
+          module: __MODULE__,
+          template: template,
+          available: unquote(template_names)
+      end
+    end
+  end
+
+  @doc false
+  def render_email(view, email, template, assigns) do
+    email
+    |> put_defaults(view)
+    |> merge_assigns(assigns)
+    |> render(template)
+  end
+
+  defp put_defaults(%{private: private} = email, view_module) do
+    private =
+      private
+      |> Map.put_new(:html_layout, false)
+      |> Map.put_new(:text_layout, false)
+      |> Map.put_new(:view_module, view_module)
+
+    %{email | private: private}
+  end
+
+  defp merge_assigns(%{assigns: email_assigns} = email, assigns) do
+    assigns = email_assigns |> Map.merge(Enum.into(assigns, %{}))
+    email |> Map.put(:assigns, assigns)
```

Comment:
```suggestion
    Map.put(email, :assigns, Map.merge(email_assigns, Enum.into(assigns, %{}))
```

Nitpick: Personally I think this suggestion is more straightforward/readable, but totally your call.

Reply:
I think I like the separate steps here. Too much going on in a single line otherwise. I don't mind the temporary `assigns` in line 104. Thanks for the suggestion though. 👍

Reply:
On second read (now that I'm looking at the full function, I like what you have there. I'll go with it.